### PR TITLE
Create a separate frontend for my geocode-postcode api

### PR DESCRIPTION
**Pull request by antoni-devlin**
This PR creates a basic frontend that allows a user to query the API, and either displays the resulting coordinates, or an error.

### CODE REVIEW

File: frontend/main.css [hunk 1]
```diff
@@ -0,0 +1,22 @@
+#map {
+	height: 70vh;
```
Comment (huwd) [inline frontend/main.css:2]: not to start another [flame war here](https://thenewstack.io/spaces-vs-tabs-a-20-year-debate-and-now-this-what-the-hell-is-wrong-with-go/) 😂

but might be worth noting GDS way prefers two spaces over tabs for both [CSS](https://gds-way.cloudapps.digital/manuals/programming-languages/css.html#whitespace) and[ javascript](https://gds-way.cloudapps.digital/manuals/programming-languages/js.html#whitespace)

it's a pedantic thing but means that if you get really deeply nested stuff or longer lines it's much easier to read.

here's a guide on how to quickly fix that in VS code

https://bobbyhadz.com/blog/vscode-change-indentation

File: frontend/main.css [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+}
+
+.spinner-loader {
+	border: 5px solid #DEE0E2;
+	border-radius: 50%;
+	border-top-color: #005EA5;
+	width: 30px;
+	height: 30px;
+	animation: spin 1s linear infinite;
```
Comment (huwd) [inline frontend/main.css:11]: oh fancy, nice use of keyframes
For a review you could also grab a screenshot of gif of that and stick it on the PR.

This looks cool but I can't visualise what it looks like without running the code otherwise!

File: frontend/script.js [hunk 3]
```diff
@@ -0,0 +1,40 @@
+async function makeRequest(postcode) {
+	const response = await fetch(`https://postcode-geocoder.onrender.com/postcode/${postcode}`);
+	const jsonResponse = await response.json();
+	if ("error" in jsonResponse) {
+		const errormsg = jsonResponse["error"];
+		return errormsg;
```
Comment (huwd) [inline frontend/script.js:6]: perhaps one for later, but you could also shorten this to be return jsonResponse["error"]

There might also be away of getting this whole block into a oneliner.
But take that as an optional improvement

**Review by huwd — COMMENTED** [review]
great, code looks good.
I think we just need to work on the commits a bit.

Perhaps `git reset main` and see if you can give a few small commits that break up logical chunks of this for a review.

Also would love to see a screenshot or animation of the final result.
Looks really exciting to have leaflet integrated!

**Comment by antoni-devlin** [comment]
GIF of the CSS spinner loader in [dc7233c](https://github.com/alphagov/learningtime-ad-sem1-postcode-geocoder/pull/47/commits/dc7233cf7d0fe9ccc9e88a1ba3cdba75a2913838). spinner-loader

File: frontend/main.css (hunk 2 above)
Comment (antoni-devlin) [reply frontend/main.css:11]: Done, added to PR! Not sure if there's a way of adding media to a PR inline, or if the way I've done it is best...

File: frontend/script.js (hunk 3 above)
Comment (antoni-devlin) [reply frontend/script.js:6]: Refactored, thanks!

File: frontend/main.css (hunk 1 above)
Comment (antoni-devlin) [reply frontend/main.css:2]: Fixed!

File: src/postcodes.ts [hunk 4]
```diff
@@ -51,24 +51,24 @@
...
 export function log(msg: any) {
-	console.log(msg);
+  console.log(msg);
 }
 
 export async function validatePostcode(postcode: string) {
-	const rawresponse = await fetch(`${base_endpoint}${postcode}/validate`);
-	const response = await rawresponse.json();
-	return response["result"];
+  const rawresponse = await fetch(`${base_endpoint}${postcode}/validate`);
```
[the remark this answers is not in the document]
Comment (Harrisman05) [reply src/postcodes.ts:58]: How did you set-up code scanning?

[the remark this answers is not in the document]
Comment (antoni-devlin) [reply src/postcodes.ts:58]: Yup! I marked this one as false positive though, Server-side request forgery shouldn't be an issue as user input is validated by Postcodes.io.

**Comment by Harrisman05** [comment]
To me, the commits nice good and atomic. A few could probably be squashed but this is probably personal preference.

As this is quite a large feature/PR relative to your project, maybe a message at the top of the PR to give a high level overview of the changes would be nice? Would be easier to follow the commit stack I think. Something like:

- Create form using HTML, CSS, JS
- Add client side script to call handle form submission and call postcodes.io to get coordinates.
- Add error handling for invalid postcode
- Implement leaflet 
- General refactor changes

Is this good advice @huwd lol??

This looks really cool with the leaflet integration, I like to see a demo of this when it's done. Nice work.

File: src/postcodes.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (antoni-devlin) [reply src/postcodes.ts:58]: Ah, misread your comment, sorry @Harrisman05 .

I turned it on in the repo settings, I think the option is called CodeQL!

[the remark this answers is not in the document]
Comment (huwd) [reply src/postcodes.ts:58]: Interesting yeah I just looked through that too.
I agree with you @antoni-devlin but let's also think about how you'd mitigate this?

You don't have to come up with a solution, but might be good to make a TODO.md or a card that outlines what you think you could do about this problem?

I wonder for instance if we could do anything to validate that input 🤔

File: src/postcodes.ts [hunk 5]
```diff
@@ -51,24 +51,24 @@
...
 }
 
 export async function geocodePostcode(postcode: string) {
-	const response = await fetch(
-		`https://api.postcodes.io/postcodes/${postcode}`
-	);
-	return await response.json();
+  const response = await fetch(
+    `https://api.postcodes.io/postcodes/${postcode}`
+  );
```
[the remark this answers is not in the document]
Comment (huwd) [reply src/postcodes.ts:66]: ditto: https://github.com/alphagov/learningtime-ad-sem1-postcode-geocoder/pull/47#discussion_r1224184004

**Comment by huwd** [comment]
Yeah I think @Harrisman05 is really onto something.

The code looks good, functionality looks fantastic.
Only thing we'd want changed here in a "real world scenario" is that reviewable unit.

So up to you at this point but the rebase stuff harley was mastering last week could be really useful here.
Or if you're not really wedded to any of your commits just a `git reset main` and then recommit.

Those bullet points Harley has suggested seem grouped around really sensible themes.

I'd perhaps look to:

- Get all those automated linting changes for whitespace into a single commit. Perhaps use that message to discuss that you've now applied linting to existing code, and this is the way going forward

Then the functional units Harley mentioned seem sensible.

Consider how you want to use your time right now, if mastering a good PR is something you are keen on. This is a good use of your time.

If you had something up next you'd rather focus on then merge and see if we can bring these principles in on the next PR (however big or small!)

**Comment by antoni-devlin** [comment]
Thanks for your suggestions! Will take into account on my next PR, but want to move the frontend work on. (Huw has said it's fine :D
